### PR TITLE
Add AXI4

### PR DESCRIPTION
**Pull request by martijnbastiaan**
TODO:

- [x] Convenience functions defined for DfLike should be exposed for the various AXI channels. This should be pretty straight-forward. It'd basically be the same as Df.
- [x] I think I made a mistake defining the Payloads (see review comments)
- [ ] It would still needs some examples (like a FIFO) on how to use it. clash-protocols is (for good reasons) a maze of type families unfortunately.
- [ ] Build an interconnect component.
- [ ] Build data generators and verifiers for functional verification.

### CODE REVIEW

File: src/Protocols/Axi4.hs [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+      -- | Write response valid
+    , _bvalid :: "BVALID" ::: Bool
+
+      -- | User data
+    , _buser :: "BUSER" ::: userType
+    }
+  deriving (Generic)
+
+data S2M_WriteResponse = S2M_WriteResponse
+  { _bready :: "BREADY" ::: Bool }
```
Comment (alex-mckenna) [inline src/Protocols/Axi4.hs:121]: :bread:

File: src/Protocols/Axi4/Raw/Full/WriteResponse.hs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    , _bvalid :: "BVALID" ::: Bool
+
+      -- | User data
+    , _buser :: "BUSER" ::: userType
+    }
+  deriving (Generic)
+
+-- | See Table A2-4 "Write response channel signals"
+data M2S_WriteResponse = S2M_WriteResponse
+  { _bready :: "BREADY" ::: Bool }
```
Comment (alex-mckenna) [inline src/Protocols/Axi4/Raw/Full/WriteResponse.hs:46]: :baguette_bread:

**Comment by jeehoonkang** [comment]
@martijnbastiaan Hi Martijn! Thank you for working on this wonderful project. I wonder if this PR is going to be merged. I really wish to use Clash for a project with AXI4 and AXI4-Stream.

(Unrelatedly, I also wonder if there is any documentation for this project. It seems [the link in the README.md](http://hackage.haskell.org/package/clash-protocols) is broken.)

File: src/Protocols/Axi4/Strict/Full/ReadAddress.hs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    C.Signal dom S2M_ReadAddress
+
+instance Backpressure (Axi4ReadAddress dom kb ksz lw iw aw kr kbl kl kc kp kq userType) where
+  boolsToBwd _ = C.fromList_lazy . coerce
+
+instance DfLike dom (Axi4ReadAddress dom kb ksz lw iw aw kr kbl kl kc kp kq) userType where
+  type Data (Axi4ReadAddress dom kb ksz lw iw aw kr kbl kl kc kp kq) userType =
+    M2S_ReadAddress kb ksz lw iw aw kr kbl kl kc kp kq userType
+
+  type Payload userType = userType
```
Comment (martijnbastiaan) [inline src/Protocols/Axi4/Strict/Full/ReadAddress.hs:63]: After working some more with AXI, I don't think exposing `userType` as the payload is a good idea. Perhaps we should define the address as the payload, so functions like `map` operate on that instead.

File: src/Protocols/Axi4/Strict/Full/ReadData.hs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    C.Signal dom M2S_ReadData
+
+instance Backpressure (Axi4ReadData dom kr iw dataType userType) where
+  boolsToBwd _ = C.fromList_lazy . coerce
+
+instance DfLike dom (Axi4ReadData dom kr iw dataType) userType where
+  type Data (Axi4ReadData dom kr iw dataType) userType =
+    S2M_ReadData kr iw dataType userType
+
+  type Payload userType = userType
```
Comment (martijnbastiaan) [inline src/Protocols/Axi4/Strict/Full/ReadData.hs:55]: Same comments as for `ReadAddress`. `dataType` probably makes more sense.

**Comment by martijnbastiaan** [comment]
Hi @jeehoonkang, thanks for the kind words :). 

There's still some things missing in this PR:

 * Convenience functions defined for `DfLike` should be exposed for the various Axi channels. This should be pretty straight-forward. It'd basically be the same as `Df`:
 
https://github.com/clash-lang/clash-protocols/blob/9795f1480f9919d9eb431d2d9f7760d5bb93c643/src/Protocols/Df.hs#L199-L201

 * I think I made a mistake defining the `Payload`s (see review comments)
 * It would still needs some examples on how to use it. `clash-protocols` is (for good reasons) a maze of type families unfortunately.

**Comment by martijnbastiaan** [comment]
> (Unrelatedly, I also wonder if there is any documentation for this project. It seems the link in the README.md is broken.)

I haven't uploaded the package to Hackage yet as I feel it isn't ready yet. The documentation is built on CI though, but it takes a bit of clicking (click green tick next to commit => details => haddock => browse job artifacts => docs => index.html). The rendered documentation for this PR can be found [here](https://clash-lang.gitlab.io/-/clash-protocols/-/jobs/945713910/artifacts/docs/index.html).

**Comment by martijnbastiaan** [comment]
Also, AXI-Stream and AXI-Lite should definitely be added! Full-blown AXI is almost always overkill! :)

I'm on another project that's draining all my energy for the next couple of months, so I doubt I'm going to have the energy to finish the PR :(.

**Comment by PietPtr** [comment]
Convenience functions have only been defined for the read data channel, as that is the only channel for which a field of kind `Type` is available (other channels do have the field `userType`, but the specification actively discourages using that field). 

Designing the FIFO is left as issue #23.

Designing the interconnect is left as issue #24.

**Comment by martijnbastiaan** [comment]
Thanks @PietPtr !

**Pull request merged**